graphium/trainer/predictor.py

```diff
@@ -449,43 +447,56 @@ def flag_step(self, batch: Dict[str, Tensor], step_name: str, to_cpu: bool) -> D
-            self.task_epoch_summary.update_predictor_state(
-                step_name="train",
-                targets=outputs["targets"],
-                predictions=outputs["preds"],
-                loss=outputs["loss"],  # This is the weighted loss for now, but change to task-specific loss
-                task_losses=outputs["task_losses"],
-                n_epochs=self.current_epoch,
-            )
-            metrics_logs = self.task_epoch_summary.get_metrics_logs()  # Dict[task, metric_logs]
-            metrics_logs["_global"]["grad_norm"] = self.get_gradient_norm()
-            outputs.update(metrics_logs)  # Dict[task, metric_logs]. Concatenate them?
 
+        # Get the metrics that are logged at every step (loss, grad_norm, batch_time, batch_tput)
         concatenated_metrics_logs = {}  # self.task_epoch_summary.concatenate_metrics_logs(metrics_logs)
         concatenated_metrics_logs["train/loss"] = outputs["loss"]
         outputs["grad_norm"] = self.get_gradient_norm()
         concatenated_metrics_logs["train/grad_norm"] = outputs["grad_norm"]
         concatenated_metrics_logs["train/batch_time"] = train_batch_time
         concatenated_metrics_logs["train/batch_tput"] = tput
+
         # report the training loss for each individual tasks
         for task in self.tasks:
             concatenated_metrics_logs[f"train/loss/{task}"] = outputs["task_losses"][task]
+
         # get the mean loss value for individual tasks as they are a tensor of size --> gradient accumulation * replication * device_iter
         for key in concatenated_metrics_logs:
             if isinstance(concatenated_metrics_logs[key], torch.Tensor):
                 if concatenated_metrics_logs[key].numel() > 1:
                     concatenated_metrics_logs[key] = concatenated_metrics_logs[key].mean()
+
+        # If logging is skipped for this step, then log the important metrics anyway and return
+        if self.skip_log_train_metrics:
+            if self.logger is not None:
+                self.logger.log_metrics(
+                    concatenated_metrics_logs, step=self.global_step
+                )  # This is a pytorch lightning function call
+            return
+
+        # this code is likely repeated for validation and testing, this should be moved to a function
+        self.task_epoch_summary.update_predictor_state(
+            step_name="train",
+            targets=outputs["targets"],
+            predictions=outputs["preds"],
+            loss=outputs["loss"],  # This is the weighted loss for now, but change to task-specific loss
+            task_losses=outputs["task_losses"],
+            n_epochs=self.current_epoch,
+        )
+        metrics_logs = self.task_epoch_summary.get_metrics_logs()  # Dict[task, metric_logs]
+        metrics_logs["_global"]["grad_norm"] = self.get_gradient_norm()
+        outputs.update(metrics_logs)  # Dict[task, metric_logs]. Concatenate them?
+
```

Comment:
It seems that `outputs` is never used anywhere but also that `metrics_logs` never makes it to `concatenated_metrics_logs`, hence the extended metrics don't appear on `wandb`. @zhiyil-graphcore can you confirm this on your side?

My fix would be to simply remove the `outputs.update` call and replace with `concatenate_metrics_logs.update(metrics_logs)`. This works on my side at least to log the extended metrics every `n` steps but the main metrics at every step.

Reply:
If you agree I will push my change and approve.

Reply:
I turned off most metrics during training so didn't catch that point. Feel free to push the changes and I can verify the throughput on IPU.

Reply:
I have made some changes, so there might be some conflicts, please beware